File: src/client/ora/orareader.cpp
/*
   Drawpile - a collaborative drawing program.

   Copyright (C) 2009-2017 Calle Laakkonen

   Drawpile is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Drawpile is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with Drawpile.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "core/blendmodes.h"
#include "core/annotationmodel.h"
#include "ora/orareader.h"
#include "ora/orawriter.h"

#include "../shared/net/layer.h"
#include "../shared/net/annotation.h"
#include "../shared/net/meta2.h"
#include "net/commands.h"
#include "utils/archive.h"
#include "utils/images.h"

#include <QGuiApplication>
#include <QImage>
#include <QXmlStreamReader>
#include <QDebug>
#include <QColor>
#include <QFile>

#include <KZip>

using protocol::MessagePtr;

namespace openraster {

namespace {
	struct Annotation {
		QRect rect;
		QColor bg;
		QString valign;
		QString content;
	};

	struct Layer {
		QString name;
		QString src;
		QPoint offset;
		qreal opacity;
		bool visibility;
		bool locked;
		QString compositeOp;
	};

	struct Canvas {
		QString error;

		QSize size;
		QList<Layer> layers; // note: layer order in an ORA stack is topmost first
		QList<Annotation> annotations;

		bool nestedWarning;
		bool extensionsWarning;

		Canvas() : nestedWarning(false), extensionsWarning(false) { }
	};
}

static bool checkIsOraFile(KArchive &zip)
{
	const QByteArray expected = "image/openraster";
	QByteArray mimetype = utils::getArchiveFile(zip, "mimetype", expected.length());

	return mimetype == expected;
}

static void skipElement(QXmlStreamReader &reader)
{
	int stack=1;
	while(!reader.atEnd() && stack>0) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::StartElement: ++stack; break;
			case QXmlStreamReader::EndElement: --stack; break;
			default: break;
		}
	}
}

/**
 * Check if the given list of attributes anything that isn't in the list of known keys
 * @param attrs attributes to check
 * @param names list of known names
 * @return true if unknown attributes are present
 */
static bool hasUnknownAttributes(const char *element, const QXmlStreamAttributes &attrs, const char **names) {
	for(const QXmlStreamAttribute &attr : attrs) {
		const char **name = names;
		bool found = false;
		while(*name) {
			if(attr.name() == *name) {
				found = true;
				break;
			}
			++name;
		}
		if(!found) {
			qWarning() << "Unknown" << element << "attribute:" << attr.name();
			return true;
		}
	}
	return false;
}

static QString attrToString(const QStringRef &attr, const QString &def)
{
	return attr.isEmpty() ? def : attr.toString();
}

static qreal attrToReal(const QStringRef &attr, double def)
{
	bool ok;
	const qreal v = attr.toDouble(&ok);
	return ok ? v : def;
}

static bool attrToBool(const QStringRef &attr, bool def, const char *trueVal)
{
	if(attr.isEmpty())
		return def;
	return attr == trueVal;
}

static QColor attrToColor(const QStringRef &attr, const QColor &def)
{
	if(attr.length() == 7 && attr.at(0) == '#')
		return QColor::fromRgb(attr.mid(1).toUInt(0, 16));
	else if(attr.length() == 9 && attr.at(0) == '#')
		return QColor::fromRgba(attr.mid(1).toUInt(0, 16));
	else
		return def;
}

//! Read a layer element
static bool readStackLayer(QXmlStreamReader &reader, Canvas &canvas, const QPoint &parentOffset)
{
	// Grab <layer> element attributes first
	const QXmlStreamAttributes attrs = reader.attributes();

	static const char *knownLayerAttributes[] = {"x", "y", "name", "src", "opacity", "visibility", "composite-op", "selected", "edit-locked", nullptr };

	if(hasUnknownAttributes("layer", attrs, knownLayerAttributes))
		canvas.extensionsWarning = true;

	Layer layer {
		attrToString(attrs.value("name"), QString()),
		attrToString(attrs.value("src"), QString()),
		parentOffset + QPoint(attrs.value("x").toInt(), attrs.value("y").toInt()),
		qBound(0.0, attrToReal(attrs.value("opacity"), 1.0), 1.0),
		attrToBool(attrs.value("visibility"), true, "visible"),
		attrToBool(attrs.value("edit-locked"), false, "true"),
		attrToString(attrs.value("composite-op"), QStringLiteral("src-over"))
	};

	canvas.layers << layer;

	// <layer> has no subelements
	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::Invalid:
				canvas.error = reader.errorString();
				return false;
			case QXmlStreamReader::StartElement:
				canvas.extensionsWarning = true;
				qWarning() << "Encountered unexpected <layer> subelement:" << reader.name();
				skipElement(reader);
				break;
			case QXmlStreamReader::EndElement:
				return true;
			default: break;
		}
	}
	
	canvas.error = "File ended unexpectedly";
	return false;
}

//! Read a drawpile specific annotation element
static bool readStackAnnotation(QXmlStreamReader &reader, Canvas &canvas)
{
	static const char *knownAttributes[] = {"x", "y", "w", "h", "valign", "bg", nullptr};
	const QXmlStreamAttributes attrs = reader.attributes();
	if(hasUnknownAttributes("annotation", attrs, knownAttributes))
		canvas.extensionsWarning = true;

	Annotation ann {
		QRect(
			attrs.value("x").toInt(),
			attrs.value("y").toInt(),
			attrs.value("w").toInt(),
			attrs.value("h").toInt()
			),
		attrToColor(attrs.value("bg"), Qt::transparent),
		attrToString(attrs.value("valign"), QString()),
		QString()
	};

	// Process <annotation> subelements
	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::Invalid:
				canvas.error = reader.errorString();
				return false;
			case QXmlStreamReader::Characters:
				ann.content += reader.text();
				break;
			case QXmlStreamReader::StartElement:
				canvas.extensionsWarning = true;
				qWarning() << "Encountered unknown <a> subelement:" << reader.name();
				skipElement(reader);
				break;
			case QXmlStreamReader::EndElement:
				canvas.annotations << ann;
				return true;
			default: break;
		}
	}

	canvas.error = "File ended unexpectedly";
	return false;
}

//! Read the drawpile specific <annotations> container element
static bool readStackAnnotations(QXmlStreamReader &reader, Canvas &canvas)
{
	static const char *knownAttributes[] = {nullptr};
	if(hasUnknownAttributes("annotations", reader.attributes(), knownAttributes))
		canvas.extensionsWarning = true;

	// Process <annotations> subelements
	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::Invalid:
				canvas.error = reader.errorString();
				return false;
			case QXmlStreamReader::StartElement:
				if(reader.name() == "a" && reader.namespaceUri() == DP_NAMESPACE) {
					if(!readStackAnnotation(reader, canvas))
						return false;

				} else {
					canvas.extensionsWarning = true;
					qWarning() << "Encountered unknown <annotations> subelement:" << reader.name();
					skipElement(reader);
				}
				break;
			case QXmlStreamReader::EndElement:
				return true;
			default: break;
		}
	}

	canvas.error = "File ended unexpectedly";
	return false;
}


//! Read a <stack> element (contain sub-stacks and layers)
static bool readStackStack(QXmlStreamReader &reader, Canvas &canvas, const QPoint &parentOffset)
{
	// Grab <stack> element attributes first
	const QXmlStreamAttributes attrs = reader.attributes();
	const QPoint offset = parentOffset + QPoint(
			attrs.value("x").toInt(),
			attrs.value("y").toInt()
			);

	static const char *knownStackAttributes[] = {"x", "y", nullptr};

	if(hasUnknownAttributes("stack", attrs, knownStackAttributes))
		canvas.extensionsWarning = true;

	// Process <stack> subelements
	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::Invalid:
				canvas.error = reader.errorString();
				return false;
			case QXmlStreamReader::StartElement:
				if(reader.name() == "stack") {
					canvas.nestedWarning = true;
					if(!readStackStack(reader, canvas, offset))
						return false;

				} else if(reader.name() == "annotations" && reader.namespaceUri() == DP_NAMESPACE) {
					if(!readStackAnnotations(reader, canvas))
						return false;

				} else if(reader.name() == "layer") {
					if(!readStackLayer(reader, canvas, QPoint()))
						return false;

				} else {
					canvas.extensionsWarning = true;
					qWarning() << "Encountered unknown <stack> subelement:" << reader.name();
					skipElement(reader);
				}
				break;
			case QXmlStreamReader::EndElement:
				return true;
			default: break;
		}
	}

	canvas.error = "File ended unexpectedly";
	return false;
}

//! Read the stack.xml root <image> element
static bool readStackImage(QXmlStreamReader &reader, Canvas &canvas)
{
	// Grab the <image> element attributes first
	const QXmlStreamAttributes attrs = reader.attributes();
	canvas.size = QSize(
			attrs.value("w").toInt(),
			attrs.value("h").toInt()
			);

	static const char *knownImageAttributes[] = {"w", "h", "version", nullptr};

	if(hasUnknownAttributes("image", attrs, knownImageAttributes))
		canvas.extensionsWarning = true;

	// Process <image> subelements
	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
			case QXmlStreamReader::Invalid:
				canvas.error = reader.errorString();
				return false;
			case QXmlStreamReader::StartElement:
				if(reader.name() == "stack") {
					if(!readStackStack(reader, canvas, QPoint()))
						return false;
				} else {
					canvas.extensionsWarning = true;
					qWarning() << "Encountered unknown <stack> subelement:" << reader.name();
					skipElement(reader);
				}
				break;
			case QXmlStreamReader::EndElement:
				return true;
			default: break;

		}
	}

	canvas.error = "File ended unexpectedly";
	return false;
}

/**
 * Generate the initialization commands from the layer stack and layer content images.
 */
static OraResult makeInitCommands(KZip &zip, const Canvas &canvas)
{
	if(canvas.size.isEmpty())
		return QGuiApplication::tr("Image has zero size!");

	if(!utils::checkImageSize(canvas.size))
		return QGuiApplication::tr("Image is too big!");

	if(canvas.layers.isEmpty())
		return QGuiApplication::tr("No layers found!");

	OraResult result;
	if(canvas.extensionsWarning)
		result.warnings |= OraResult::ORA_EXTENDED;
	if(canvas.nestedWarning)
		result.warnings |= OraResult::ORA_NESTED;

	const int ctxId = 1;

	// Set canvas size
	result.commands << MessagePtr(new protocol::CanvasResize(ctxId, 0, canvas.size.width(), canvas.size.height(), 0));

	// Create layers
	// Note: layers are stored topmost first in ORA, but we create them bottom-most first
	int layerId = ctxId << 8;
	for(int i=canvas.layers.size()-1;i>=0;--i) {
		const Layer &layer = canvas.layers[i];
		QImage content;
		{
			QByteArray image = utils::getArchiveFile(zip, layer.src);
			if(image.isNull() || !content.loadFromData(image)) {
				return QGuiApplication::tr("Couldn't load layer %1").arg(layer.src);
			}
		}

		const QColor solidColor = utils::isSolidColorImage(content);

		result.commands.append(MessagePtr(new protocol::LayerCreate(
				ctxId,
				++layerId,
				0,
				solidColor.isValid() ? solidColor.rgba() : 0,
				0,
				layer.name
			)));

		if(!solidColor.isValid())
			result.commands << net::command::putQImage(
					ctxId,
					layerId,
					layer.offset.x(),
					layer.offset.y(),
					content,
					paintcore::BlendMode::MODE_REPLACE
					);

		bool exact_blendop;
		int blendmode = paintcore::findBlendModeByName(layer.compositeOp, &exact_blendop).id;
		if(!exact_blendop)
			result.warnings |= OraResult::ORA_EXTENDED;

		result.commands << MessagePtr(new protocol::LayerAttributes(
			ctxId,
			layerId,
			qRound(255 * layer.opacity),
			blendmode
			));

		if(layer.locked) {
			result.commands << MessagePtr(new protocol::LayerACL(ctxId, layerId, true, QList<uint8_t>()));
		}

		if(!layer.visibility) {
			result.commands << MessagePtr(new protocol::LayerVisibility(ctxId, layerId, false));
		}
	}

	// Create annotations
	int annotationId = ctxId << 8;
	for(const Annotation &ann : canvas.annotations) {
		result.commands.append(MessagePtr(new protocol::AnnotationCreate(
			ctxId,
			++annotationId,
			ann.rect.x(),
			ann.rect.y(),
			ann.rect.width(),
			ann.rect.height()
		)));
		result.commands.append(MessagePtr(new protocol::AnnotationEdit(
			ctxId,
			annotationId,
			ann.bg.rgba(),
			paintcore::Annotation::valignFromString(ann.valign),
			0,
			ann.content
		)));
	}

	return result;
}

OraResult loadOpenRaster(const QString &filename)
{
	Canvas canvas;
	QFile orafile(filename);
	KZip zip(&orafile);

	if(!zip.open(QIODevice::ReadOnly)) {
		return orafile.errorString();
	}

	// Make sure this is an OpenRaster file
	if(!checkIsOraFile(zip)) {
		return QGuiApplication::tr("File is not an OpenRaster file");
	}

	// Read the layer stack definition
	QXmlStreamReader reader(zip.directory()->file("stack.xml")->data());

	while(!reader.atEnd()) {
		const auto tokentype = reader.readNext();
		switch(tokentype) {
		case QXmlStreamReader::Invalid:
			return reader.errorString();
		case QXmlStreamReader::StartElement:
			if(reader.name() == "image") {
				if(!readStackImage(reader, canvas))
					return canvas.error;
			} else {
				return QStringLiteral("Unexpected root element: ") + reader.name().toString();
			}
		default: break;
		}
	}

	// Generate the commands to initialize session history
	return makeInitCommands(zip, canvas);
}

}

File: src/dpimpex/src/ora_reader.rs
// This file is part of Drawpile.
// Copyright (C) 2021 Calle Laakkonen
//
// Drawpile is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// As additional permission under section 7, you are allowed to distribute
// the software through an app store, even if that store has restrictive
// terms and conditions that are incompatible with the GPL, provided that
// the source is also available under the GPL with or without this permission
// through a channel without those restrictive terms and conditions.
//
// Drawpile is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Drawpile.  If not, see <https://www.gnu.org/licenses/>.

use super::conv::to_dpimage;
use super::ora_utils::{DP_NAMESPACE, MYPAINT_NAMESPACE};
use super::{ImageImportResult, ImpexError};
use dpcore::paint::annotation::VAlign;
use dpcore::paint::layerstack::{LayerFill, LayerInsertion};
use dpcore::paint::{editlayer, Blendmode, Color, Image, LayerStack, Rectangle, Size, Tile};

use std::fs::File;
use std::io::{Cursor, Read, Seek};
use std::path::Path;
use std::str::FromStr;

use image::io::Reader as ImageReader;
use tracing::warn;
use xml::attribute::OwnedAttribute;
use xml::reader::{EventReader, XmlEvent};
use zip::ZipArchive;

pub fn load_openraster_image(path: &Path) -> ImageImportResult {
    let mut archive = ZipArchive::new(File::open(path)?)?;

    check_mimetype(&mut archive)?;
    let canvas = parse_stack_xml(archive.by_name("stack.xml")?)?;

    let mut ls = LayerStack::new(canvas.size.width as u32, canvas.size.height as u32);

    // TODO dpi

    // Create layers
    let mut layer_id = 0x0101;
    for oralayer in canvas.layers {
        if !oralayer.bgtile.is_empty() {
            let tile_image = get_image_file(&mut archive, &oralayer.bgtile)?;
            if tile_image.width == 64 || tile_image.height == 64 {
                // If we have a valid tile, this is a background layer
                // that we don't need to represent as an actual layer
                ls.background = Tile::from_data(&tile_image.pixels, 0);
                continue;
            }
        }

        let img = get_image_file(&mut archive, &oralayer.filename)?;

        let mut layer = ls
            .add_layer(
                layer_id,
                LayerFill::Solid(Color::TRANSPARENT),
                LayerInsertion::Top,
            )
            .unwrap();

        layer.title = format!("Layer {}", layer_id - 0x0100);
        layer.opacity = oralayer.opacity.clamp(0.0, 1.0);
        layer.hidden = !oralayer.visibility;
        layer.censored = oralayer.censored;
        layer.fixed = oralayer.fixed;
        layer.blendmode = Blendmode::from_svg_name(&oralayer.composite_op).unwrap_or_default();

        // TODO layer lock status

        editlayer::draw_image(
            &mut layer,
            1,
            &img.pixels,
            &Rectangle::new(
                oralayer.offset.0,
                oralayer.offset.1,
                img.width as i32,
                img.height as i32,
            ),
            1.0,
            Blendmode::Replace,
        );

        layer_id += 1;
    }

    // Create annotations
    let annotation_id = 0x0101;
    for ora_ann in canvas.annotations {
        ls.add_annotation(annotation_id, ora_ann.rect);
        let mut a = ls.get_annotation_mut(annotation_id).unwrap();
        a.text = ora_ann.content;
        a.background = Color::from_str(&ora_ann.bg).unwrap_or(Color::TRANSPARENT);
        a.valign = match ora_ann.valign.as_ref() {
            "center" => VAlign::Center,
            "bottom" => VAlign::Bottom,
            _ => VAlign::Top,
        };
    }

    // TODO warn about unsupported features
    Ok(ls)
}

/// OpenRaster files are identified by the presence of a "mimetype" file
fn check_mimetype<R: Read + Seek>(archive: &mut ZipArchive<R>) -> Result<(), ImpexError> {
    let expected_mimetype = b"image/openraster";

    let mut mtfile = archive.by_name("mimetype")?;
    if mtfile.size() != expected_mimetype.len() as u64 {
        return Err(ImpexError::UnsupportedFormat);
    }

    let mut mimetype = Vec::new();
    mtfile.read_to_end(&mut mimetype)?;

    if mimetype == expected_mimetype {
        Ok(())
    } else {
        Err(ImpexError::UnsupportedFormat)
    }
}

fn get_image_file<R: Read + Seek>(
    archive: &mut ZipArchive<R>,
    filename: &str,
) -> Result<Image, ImpexError> {
    let mut filecontent = Vec::new();
    archive.by_name(filename)?.read_to_end(&mut filecontent)?;
    let img = ImageReader::new(Cursor::new(filecontent))
        .with_guessed_format()?
        .decode()?;

    Ok(to_dpimage(&img.into_rgba8()))
}

#[derive(Debug)]
struct OraCanvas {
    size: Size,
    unsupported_features: bool,
    dpi: (i32, i32),
    layers: Vec<OraLayer>,           // note: nested stacks not yet supported
    annotations: Vec<OraAnnotation>, // our extension
}

#[derive(Debug)]
struct OraLayer {
    name: String,
    filename: String,
    bgtile: String, // MyPaint extension
    offset: (i32, i32),
    opacity: f32,
    visibility: bool,
    locked: bool,
    censored: bool, // our extension
    fixed: bool,    // our extension
    composite_op: String,
    unsupported_features: bool, // this layer makes use of unsupported features
}

#[derive(Debug)]
struct OraAnnotation {
    rect: Rectangle,
    bg: String,
    valign: String,
    content: String,
}

fn parse_stack_xml<R: Read>(file: R) -> Result<OraCanvas, ImpexError> {
    let mut parser = EventReader::new(file);

    // Expect <image> root element
    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement {
                name, attributes, ..
            }) => {
                if name.local_name != "image" {
                    warn!(
                        "Error reading OpenRaster stack: expected <image>, got <{}>",
                        name
                    );
                    return Err(ImpexError::UnsupportedFormat);
                }

                return parse_stack_image(attributes, &mut parser);
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster stack: Unexpected end of document");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster stack: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => {}
        }
    }
}

fn parse_stack_image<R: Read>(
    mut attributes: Vec<OwnedAttribute>,
    parser: &mut EventReader<R>,
) -> Result<OraCanvas, ImpexError> {
    let mut canvas = OraCanvas {
        size: Size::new(
            take_attribute(&mut attributes, "w", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
            take_attribute(&mut attributes, "h", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
        ),
        unsupported_features: false,
        dpi: (
            take_attribute(&mut attributes, "xres", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(72),
            take_attribute(&mut attributes, "yres", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(72),
        ),
        layers: Vec::new(),
        annotations: Vec::new(),
    };

    if canvas.size.width <= 0 || canvas.size.height <= 0 {
        warn!("OpenRaster file has invalid image size");
        return Err(ImpexError::UnsupportedFormat);
    }
    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement {
                name, attributes, ..
            }) => {
                if name.local_name == "stack" {
                    parse_stack_stack(&mut canvas, attributes, (0, 0), parser)?;
                } else {
                    warn!("Unsupported openraster <image> element <{}>", name);
                    canvas.unsupported_features = true;
                    skip_element(parser)?;
                }
            }
            Ok(XmlEvent::EndElement { name, .. }) => {
                assert_eq!(name.local_name, "image");
                // Note: layers are listed top-to-bottom in the XML,
                // but are stored bottom-to-top inside Drawpile
                canvas.layers.reverse();
                return Ok(canvas);
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster stack: Unexpected end of document while parsing <image>");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster stack: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => {}
        }
    }
}

fn parse_stack_stack<R: Read>(
    canvas: &mut OraCanvas,
    mut attributes: Vec<OwnedAttribute>,
    offset: (i32, i32),
    parser: &mut EventReader<R>,
) -> Result<(), ImpexError> {
    let offset = (
        offset.0
            + take_attribute(&mut attributes, "x", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
        offset.1
            + take_attribute(&mut attributes, "y", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
    );

    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement {
                name, attributes, ..
            }) => {
                if name.local_name == "stack" {
                    // Nested stacks aren't supported at the moment, but we can
                    // still get the layers out of  them.
                    canvas.unsupported_features = true;
                    parse_stack_stack(canvas, attributes, offset, parser)?;
                } else if name.local_name == "layer" {
                    canvas
                        .layers
                        .push(parse_stack_layer(attributes, offset, parser)?);
                } else if name.local_name == "annotations"
                    && name.namespace.as_deref() == Some(DP_NAMESPACE)
                {
                    canvas.annotations.append(&mut parse_annotations(parser)?);
                } else {
                    warn!("Unsupported OpenRaster stack element <{}>", name.local_name);
                    canvas.unsupported_features = true;
                    skip_element(parser)?;
                }
            }
            Ok(XmlEvent::EndElement { name, .. }) => {
                assert_eq!(name.local_name, "stack");
                return Ok(());
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster stack: Unexpected end of document while parsing <image>");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster stack: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => (),
        }
    }
}
fn parse_stack_layer<R: Read>(
    mut attributes: Vec<OwnedAttribute>,
    offset: (i32, i32),
    parser: &mut EventReader<R>,
) -> Result<OraLayer, ImpexError> {
    let mut layer = OraLayer {
        name: take_attribute(&mut attributes, "name", None).unwrap_or(String::new()),
        filename: take_attribute(&mut attributes, "src", None)
            .ok_or(ImpexError::UnsupportedFormat)?,
        bgtile: take_attribute(&mut attributes, "background-tile", Some(MYPAINT_NAMESPACE))
            .unwrap_or(String::new()),
        offset: (
            offset.0
                + take_attribute(&mut attributes, "x", None)
                    .and_then(|a| a.parse::<i32>().ok())
                    .unwrap_or(0),
            offset.1
                + take_attribute(&mut attributes, "y", None)
                    .and_then(|a| a.parse::<i32>().ok())
                    .unwrap_or(0),
        ),
        opacity: take_attribute(&mut attributes, "opacity", None)
            .and_then(|a| a.parse::<f32>().ok())
            .unwrap_or(1.0),
        visibility: take_attribute(&mut attributes, "visibility", None)
            .map_or(true, |a| a == "visibile"),
        locked: take_attribute(&mut attributes, "edit-locked", None).map_or(false, |a| a == "true"),
        censored: take_attribute(&mut attributes, "censored", Some(DP_NAMESPACE))
            .map_or(false, |a| a == "true"),
        fixed: take_attribute(&mut attributes, "fixed", Some(DP_NAMESPACE))
            .map_or(false, |a| a == "true"),
        composite_op: take_attribute(&mut attributes, "composite-op", None)
            .unwrap_or("src-over".into()),
        unsupported_features: false,
    };

    if !attributes.is_empty() {
        warn!("Unsupported OpenRaster layer attributes: {:?}", attributes);
        layer.unsupported_features = true;
    }

    // Layer element shouldn't have any child elements
    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement { name, .. }) => {
                warn!("Unsupported OpenRaster layer element <{}>", name.local_name);
                layer.unsupported_features = true;
                skip_element(parser)?;
            }
            Ok(XmlEvent::EndElement { name, .. }) => {
                assert_eq!(name.local_name, "layer");
                return Ok(layer);
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster stack: Unexpected end of document while parsing <image>");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster stack: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => (),
        }
    }
}

fn parse_annotations<R: Read>(
    parser: &mut EventReader<R>,
) -> Result<Vec<OraAnnotation>, ImpexError> {
    let mut annotations = Vec::new();

    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement {
                name, attributes, ..
            }) => {
                if name.local_name == "a" {
                    annotations.push(parse_annotation(attributes, parser)?);
                } else {
                    warn!(
                        "Unsupported OpenRaster annotations element <{}>",
                        name.local_name
                    );
                    skip_element(parser)?;
                }
            }
            Ok(XmlEvent::EndElement { name, .. }) => {
                assert_eq!(name.local_name, "annotations");
                return Ok(annotations);
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster annotations: Unexpected end of document while parsing <image>");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster annotations: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => (),
        }
    }
}

fn parse_annotation<R: Read>(
    mut attributes: Vec<OwnedAttribute>,
    parser: &mut EventReader<R>,
) -> Result<OraAnnotation, ImpexError> {
    let mut annotation = OraAnnotation {
        rect: Rectangle::new(
            take_attribute(&mut attributes, "x", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
            take_attribute(&mut attributes, "y", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
            take_attribute(&mut attributes, "w", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
            take_attribute(&mut attributes, "h", None)
                .and_then(|a| a.parse::<i32>().ok())
                .unwrap_or(0),
        ),
        bg: take_attribute(&mut attributes, "bg", None).unwrap_or(String::new()),
        valign: take_attribute(&mut attributes, "valign", None).unwrap_or(String::new()),
        content: String::new(),
    };

    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement { name, .. }) => {
                warn!(
                    "Unsupported OpenRaster annotation element <{}>",
                    name.local_name
                );
                skip_element(parser)?;
            }
            Ok(XmlEvent::EndElement { name, .. }) => {
                assert_eq!(name.local_name, "a");
                return Ok(annotation);
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster annotation: Unexpected end of document while parsing <image>");
                return Err(ImpexError::UnsupportedFormat);
            }
            Ok(XmlEvent::CData(s)) => annotation.content.push_str(&s),
            Ok(XmlEvent::Characters(s)) => annotation.content.push_str(&s),
            Err(e) => {
                warn!("Error reading OpenRaster annotation: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => (),
        }
    }
}

fn skip_element<R: Read>(parser: &mut EventReader<R>) -> Result<(), ImpexError> {
    let mut depth = 1;
    loop {
        match parser.next() {
            Ok(XmlEvent::StartElement { .. }) => {
                depth += 1;
            }
            Ok(XmlEvent::EndElement { .. }) => {
                depth -= 1;
                if depth == 0 {
                    return Ok(());
                }
            }
            Ok(XmlEvent::EndDocument) => {
                warn!("Error reading OpenRaster stack: Unexpected end of document");
                return Err(ImpexError::UnsupportedFormat);
            }
            Err(e) => {
                warn!("Error reading OpenRaster stack: {}", e);
                return Err(ImpexError::UnsupportedFormat);
            }
            _ => (),
        }
    }
}

fn take_attribute(
    attrs: &mut Vec<OwnedAttribute>,
    name: &str,
    namespace: Option<&str>,
) -> Option<String> {
    if let Some(idx) = attrs
        .iter()
        .position(|a| a.name.local_name == name && a.name.namespace.as_deref() == namespace)
    {
        Some(attrs.remove(idx).value)
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_load_openraster() {
        let path: PathBuf = [env!("CARGO_MANIFEST_DIR"), "testdata", "oratest.ora"]
            .iter()
            .collect();
        let ls = load_openraster_image(path.as_ref()).unwrap();
        let annotations = ls.get_annotations();

        assert_eq!(ls.layer_count(), 3);
        assert_eq!(annotations.len(), 1);

        assert!(matches!(annotations[0].valign, VAlign::Center));
    }
}
